mcstatus/tests/test_address.py

```diff
@@ -0,0 +1,216 @@
+            address = minecraft_srv_address_lookup("example.org", default_port=25565, lifetime=3)
+            resolve.assert_called_once_with("_minecraft._tcp.example.org", RdataType.SRV, lifetime=3)
+
+        assert address.host == "example.org"
+        assert address.port == 25565
+
+    def test_address_with_srv(self):
+        with patch("dns.resolver.resolve") as resolve:
+            answer = Mock()
+            answer.target = "different.example.org."
+            answer.port = 12345
+            resolve.return_value = [answer]
+
+            address = minecraft_srv_address_lookup("example.org", lifetime=3)
+            resolve.assert_called_once_with("_minecraft._tcp.example.org", RdataType.SRV, lifetime=3)
+        assert address.host == "different.example.org"
+        assert address.port == 12345
+
+    @pytest.mark.asyncio
+    async def test_async_address_no_srv(self):
+        with patch("dns.asyncresolver.resolve") as resolve:
+            resolve.side_effect = [dns.resolver.NXDOMAIN]
+            address = await async_minecraft_srv_address_lookup("example.org", default_port=25565, lifetime=3)
+            resolve.assert_called_once_with("_minecraft._tcp.example.org", RdataType.SRV, lifetime=3)
+
+        assert address.host == "example.org"
+        assert address.port == 25565
+
+    @pytest.mark.asyncio
+    async def test_async_address_with_srv(self):
+        with patch("dns.asyncresolver.resolve") as resolve:
+            answer = Mock()
+            answer.target = "different.example.org."
+            answer.port = 12345
+            resolve.side_effect = const_coro([answer])
+
+            address = await async_minecraft_srv_address_lookup("example.org", lifetime=3)
+            resolve.assert_called_once_with("_minecraft._tcp.example.org", RdataType.SRV, lifetime=3)
+        assert address.host == "different.example.org"
+        assert address.port == 12345
+
+
+class TestAddressValidity:
+    def test_address_validation_valid(self):
+        Address._ensure_validity("example.org", 25565)
+        Address._ensure_validity("192.168.0.100", 54321)
+        Address._ensure_validity("2345:0425:2CA1:0000:0000:0567:5673:23b5", 100)
+        Address._ensure_validity("2345:0425:2CA1::0567:5673:23b5", 12345)
+
+    def test_address_validation_invalid_port(self):
+        # Shouldn't accept port out of range
+        with pytest.raises(ValueError):
+            Address._ensure_validity("example.org", 100_000)
+        with pytest.raises(ValueError):
+            Address._ensure_validity("example.org", -1)
+
+    def test_address_validation_invalid_types(self):
+        cases = (
+            ("example.org", "25565"),
+            (25565, "example.org"),
+            (("example.org", 25565), None),
+            (0, 0),
+            ("", ""),
+        )
+        for test_host, test_port in cases:
+            with pytest.raises(TypeError):
+                Address._ensure_validity(test_host, test_port)
```

Comment:
Pytest offers a marker for running the same test multiple times with different inputs https://docs.pytest.org/en/latest/how-to/parametrize.html

This would look something like
```suggestion
    @pytest.mark.parametrize(
        "test_host,test_port,expected_exception",
        [
            ("example.org", 100_000, ValueError),
            ("example.org", -1, ValueError),
            ("example.org", "25565", TypeError),
            (25565, "example.org", TypeError),
            (("example.org", 25565), None, TypeError),
            (0, 0, TypeError),
            ("", "", TypeError),
        ],
    )
    def test_address_validation_invalid_port(self, test_host, test_port, expected_exception):
        with pytest.raises(expected_exception):
            Address._ensure_validity(test_host, test_port)
```

Reply:
Ah, nice, I was wondering how to implement multiple cases without multiple functions in pytest, I'm more used to pure unittest testing, thanks!